Feat: Add media output switcher support for original equipment manufacturers (OEMs)

ghhccghk opened the pull request:
Code sources: https://github.com/Moriafly/media-kit and https://github.com/Yos-X/FlamingoSank 

The functionality works as follows across different Android versions and OEM ROMs:
- For Android 14 and higher, it uses MediaRouter2#showSystemOutputSwitcher.
- For Android 12 and higher, it uses the “com.android.systemui.action.LAUNCH_MEDIA_OUTPUT_DIALOG” action.
- For Android 11, it uses the “com.android.settings.panel.action.MEDIA_OUTPUT” action.
- For MIUI (Xiaomi), it uses “miui.systemui.miplay.MiPlayDetailActivity”.
- For One UI (Samsung), it uses “com.samsung.android.mdx.quickboard.view.MediaActivity”.

ghhccghk committed:
Feat: Add media output switcher support for original equipment manufacturers (OEMs)
Code sources: https://github.com/Moriafly/media-kit and https://github.com/Yos-X/FlamingoSank
The functionality works as follows across different Android versions and OEM ROMs:
- For Android 14 and higher, it uses MediaRouter2#showSystemOutputSwitcher.
- For Android 12 and higher, it uses the “com.android.systemui.action.LAUNCH_MEDIA_OUTPUT_DIALOG” action.
- For Android 11, it uses the “com.android.settings.panel.action.MEDIA_OUTPUT” action.
- For MIUI (Xiaomi), it uses “miui.systemui.miplay.MiPlayDetailActivity”.
- For One UI (Samsung), it uses “com.samsung.android.mdx.quickboard.view.MediaActivity”.

ghhccghk committed:
Feat: Add media output switcher support for original equipment manufacturers (OEMs)
Code sources: https://github.com/Moriafly/media-kit and https://github.com/Yos-X/FlamingoSank
The functionality works as follows across different Android versions and OEM ROMs:
- For Android 14 and higher, it uses MediaRouter2#showSystemOutputSwitcher.
- For Android 12 and higher, it uses the “com.android.systemui.action.LAUNCH_MEDIA_OUTPUT_DIALOG” action.
- For Android 11, it uses the “com.android.settings.panel.action.MEDIA_OUTPUT” action.
- For MIUI (Xiaomi), it uses “miui.systemui.miplay.MiPlayDetailActivity”.
- For One UI (Samsung), it uses “com.samsung.android.mdx.quickboard.view.MediaActivity”.

ghhccghk committed:
Fix: Refactored the SystemMediaControlResolver

The SystemMediaControlResolver has been refactored to better handle various Android versions and OEM customizations. It also introduces dedicated checks for One UI versions and displays a Toast message when media output settings cannot be opened.

Specific changes include:

* The logic for launching the system media control dialog has been refactored into a new private function `startSystemMediaControl()`.
* Added a new function `getOneUIVersionReadable()` to retrieve the One UI version.
* Fall back to `startSystemMediaControl()` when launching OEM-specific media control dialogs fails.
* Display a Toast message when media output settings cannot be opened.
* Added new strings in `strings.xml`.

ghhccghk committed:
Fix: Refactored the SystemMediaControlResolver

The SystemMediaControlResolver has been refactored to better handle various Android versions and OEM customizations. It also introduces dedicated checks for One UI versions and displays a Toast message when media output settings cannot be opened.

Specific changes include:

* The logic for launching the system media control dialog has been refactored into a new private function `startSystemMediaControl()`.
* Added a new function `getOneUIVersionReadable()` to retrieve the One UI version.
* Fall back to `startSystemMediaControl()` when launching OEM-specific media control dialogs fails.
* Display a Toast message when media output settings cannot be opened.
* Added new strings in `strings.xml`.

ghhccghk committed:
Refactor: Improve One UI version detection

Moved the `getSystemProperties` logic into a separate private function for better readability and potential reuse. This change also ensures that an `IllegalArgumentException` is properly thrown if encountered during reflection.

nift4 commented on app/src/main/java/org/akanework/gramophone/logic/utils/exoplayer/oem/SystemMediaControlResolver.kt:
```diff
@@ -0,0 +1,166 @@
+            if (!tag) {
+                Toast.makeText(context, R.string.media_control_text_error, Toast.LENGTH_SHORT).show()
+            }
+        }
```
isn't this duplicated?

nift4 replied:
android 11 and below code appears to be duplicated @ghhccghk

ghhccghk commented:
If you have any more questions, let's talk about them tomorrow. I'm going to bed now.

ghhccghk committed:
fix：Merge code for calling media controls on Android 11 and lower versions

nift4 commented:
@ghhccghk 

Can you please:
- move the output switcher button to the top right like you mentioned before
- use .xml icon for output switcher instead of .png
- remove xiaomi and samsung specific code for now, because it'll get stale and there's no way for me to know when xiaomi or samsung update their apis
- use "git rebase" to update your pr branch on top of latest gramophone

thank you!

ghhccghk commented:
> @ghhccghk
> 
> Can you please:
> 
> * move the output switcher button to the top right like you mentioned before
> * use .xml icon for output switcher instead of .png
> * remove xiaomi and samsung specific code for now, because it'll get stale and there's no way for me to know when xiaomi or samsung update their apis
> * use "git rebase" to update your pr branch on top of latest gramophone
> 
> thank you!

OK, I have confirmed that Xiaomi will not remove the api, because this is one of their important interfaces for interconnection.

ghhccghk committed:
Modify the position of the media control buttons in the full_player.xml layout.

The media control buttons have been removed from the linear layout at the bottom of the screen and moved to the top of the screen. This button is now fixed at the top of the parent layout and has some margin for adjustment.

ghhccghk committed:
Remove One UI media activity calls

Temporarily disable One UI MediaActivity calls and wait until it is certain that they will not be deleted before adding them back.

Signed-off-by: ghhccghk <2137610394@qq.com>

nift4 commented:
So, the current problem is that it will show button for old device which never had output switcher. Like Android 5, 6, 7, 8, etc.

ghhccghk commented:
> So, the current problem is that it will show button for old device which never had output switcher. Like Android 5, 6, 7, 8, etc.

Then add detection, and hide it if it is not supported

nift4 commented:
Sure, I'll do it soon.

ghhccghk commented:
> Sure, I'll do it soon.

OK

ghhccghk committed:
feat: Optimized media output panel availability checks

Adjusted media output panel availability check logic based on Android version.

- For Android 14 and above, assume the panel is always available.
- For Android 12 and 13, check if the system UI can handle the media output dialog intent.
- For Android 11 and below, check if the Settings app can handle the media output panel intent.

If the panel is unavailable, hide the media control buttons.

Signed-off-by: ghhccghk <2137610394@qq.com>

ghhccghk committed:
feat: Optimized media output panel availability checks

Adjusted media output panel availability check logic based on Android version.

- For Android 14 and above, assume the panel is always available.
- For Android 12 and 13, check if the system UI can handle the media output dialog intent.
- For Android 11 and below, check if the Settings app can handle the media output panel intent.

If the panel is unavailable, hide the media control buttons.

Signed-off-by: ghhccghk <2137610394@qq.com>

nift4 commented:
Hi, just to clarify what's the status: I think I will merge this after next release when I have more time. Reason is that 1. this is good opportunity to move more buttons around and migrate to proper TopAppBar in that screen, 2. as fallback I want to implement in-app output switcher (for A10 and earlier), 3. I want to check if it's possible to make MediaRouter allow to change between headphones and speaker before A15 in this output switcher, 4. for the previous point it would be good if Google fixed https://issuetracker.google.com/issues/340805981

Basically I want to check how much potential MediaRouter has before merging this feature, but I promise I will get back to you. Thank you for your patience!

ghhccghk commented:
> Hi, just to clarify what's the status: I think I will merge this after next release when I have more time. Reason is that 1. this is good opportunity to move more buttons around and migrate to proper TopAppBar in that screen, 2. as fallback I want to implement in-app output switcher (for A10 and earlier), 3. I want to check if it's possible to make MediaRouter allow to change between headphones and speaker before A15 in this output switcher, 4. for the previous point it would be good if Google fixed https://issuetracker.google.com/issues/340805981
> 
> Basically I want to check how much potential MediaRouter has before merging this feature, but I promise I will get back to you. Thank you for your patience!

no problem

Commit:
feat: Add playback speed control

Adds a new button and dialog to the player which allows you to set a custom playback speed

Commit:
Move playback speed button where requested

ghhccghk committed:
I addressed the differences between the media control buttons and the main code.

Signed-off-by: ghhccghk <2137610394@qq.com>

ghhccghk committed:
fix: Addressing issues caused by variable base code

Signed-off-by: ghhccghk <2137610394@qq.com>

ghhccghk commented:
I have completed the code merge and tested that it works properly. Awaiting review and merge.

nift4 commented:
Hi @ghhccghk,

I found some issue that should be addressed:
- in landscape mode, it crashes. (because res/values-land/full_player.xml was not updated)
- the output switch button should be next to speed change button at the top of the screen
- the output switch button uses .png icon, but it should use .xml icon

ghhccghk committed:
fix：Move the media control buttons to the top bar

Signed-off-by: ghhccghk <2137610394@qq.com>

ghhccghk committed:
feat: Converted media control icon to vector graphics

Removed the PNG version of `ic_media_control.png` and replaced it with a vector (XML) version. This helps reduce app size and improves scaling quality across different screen densities.

Signed-off-by: ghhccghk <2137610394@qq.com>

ghhccghk commented:
That's good enough. Take a look at the code.

> Hi @ghhccghk,
> 
> I found some issue that should be addressed:
> 
> * in landscape mode, it crashes. (because res/values-land/full_player.xml was not updated)
> * the output switch button should be next to speed change button at the top of the screen
> * the output switch button uses .png icon, but it should use .xml icon

nift4 commented on app/src/main/java/org/akanework/gramophone/logic/utils/exoplayer/oem/UnstableMediaKitApi.kt:
```diff
@@ -0,0 +1,26 @@
+    level = RequiresOptIn.Level.ERROR,
+)
+@Retention(AnnotationRetention.BINARY)
+annotation class UnstableMediaKitApi
```
please delete this file

nift4 commented on app/src/main/java/org/akanework/gramophone/ui/components/FullBottomSheet.kt:
```diff
@@ -1294,4 +1310,55 @@ class FullBottomSheet
         }
     }
 
+    fun isMediaOutputPanelSupported(context: Context): Boolean {
```
can you please move this judgement into SystemMediaControlResolver? thanks!

nift4 commented on app/src/main/java/org/akanework/gramophone/logic/utils/exoplayer/oem/SystemMediaControlResolver.kt:
```diff
@@ -0,0 +1,157 @@
+                    startSystemMediaControl()
+                }
+            }
+            //zh：临时禁用OneUI MediaActivity调用，等待未来确定不会被删除再添加回去
```
please delete commented out code

nift4 commented on app/src/main/java/org/akanework/gramophone/logic/utils/exoplayer/oem/MiPlayAudioSupport.kt:
```diff
@@ -0,0 +1,123 @@
+        }
+
+    /**
+     * 检查 SystemUI 是否包含处理妙播意图的 Activity
```
please translate all chinese comment to english

ghhccghk committed:
Optimize code by removing redundant logic

Signed-off-by: ghhccghk <2137610394@qq.com>

ghhccghk committed:
feat(UI): Refactor full player layout and build configurations

- In `build.gradle.kts`, set the debug signing configuration for the release build type.
- In `full_player.xml`, reorder the media control and playback speed buttons.
- In `layout-w600dp-land/full_player.xml`, clean up redundant layout attributes for the album cover card view.

Signed-off-by: ghhccghk <2137610394@qq.com>

ghhccghk committed:
fix merger error

Signed-off-by: ghhccghk <2137610394@qq.com>

ghhccghk committed:
remove userdebug signingConfig

Signed-off-by: ghhccghk <2137610394@qq.com>
